### PR TITLE
wr-161: Apply articles reaction flow from FE to BE

### PR DESCRIPTION
**Pull request by hnoievets**
- [x] Return likes/dislikes count with articles on GET /articles /articles/:id
- [x] Add actions to like/dislike article

Own article:
image

Others article:
image

Also:
- renamed ArticleWithAuthType to ArticleWithRelationsType
- used new type ArticleResponseDto on FE
- unified article response

### CODE REVIEW

File: backend/src/packages/articles/libs/helpers/join-article-realations.helper.ts
Comment (andanf-e) [inline backend/src/packages/articles/libs/helpers/join-article-realations.helper.ts:1]: Make it as private method in articles repository

File: backend/src/packages/articles/libs/helpers/modify-reactions-graph.helper.ts
Comment (andanf-e) [inline backend/src/packages/articles/libs/helpers/modify-reactions-graph.helper.ts:1]: Make it as private method in articles repository

File: frontend/src/slices/articles/articles.slice.ts [hunk 1]
```diff
@@ -54,6 +55,10 @@ const { reducer, actions, name } = createSlice({
       state.dataStatus = DataStatus.FULFILLED;
       state.article = action.payload;
     });
+    builder.addCase(reactToArticle.fulfilled, (state, action) => {
+      state.dataStatus = DataStatus.FULFILLED;
```
Comment (andanf-e) [inline frontend/src/slices/articles/articles.slice.ts]: Use separate dataStatus for articleReaction action

File: frontend/src/slices/articles/actions.ts [hunk 2]
```diff
@@ -79,4 +79,34 @@ const getArticle = createAsyncThunk<
...
+
+  const articleWithUpdatedReaction = await articleApi.getArticle(
+    payload.articleId,
+  );
+
+  return articles.map((article) =>
+    article.id === articleWithUpdatedReaction.id
+      ? articleWithUpdatedReaction
+      : article,
+  );
```
Comment (andanf-e) [inline frontend/src/slices/articles/actions.ts]: This action should take the reaction to insert\update and return inserted\updated - not get article and map through articles

File: frontend/src/slices/articles/articles.slice.ts [hunk 3]
```diff
@@ -54,6 +55,10 @@ const { reducer, actions, name } = createSlice({
...
+      state.articles = action.payload;
```
Comment (andanf-e) [inline frontend/src/slices/articles/articles.slice.ts]: Move articles update upon reaction action here + take separately reaction and articleId from action payload

File: frontend/src/pages/articles/components/article-card/article-card.tsx [hunk 4]
```diff
@@ -91,21 +119,40 @@ const ArticleCard: React.FC<Properties> = ({
...
+          <li>
+            <IconButton
+              iconName="like"
+              className={getValidClassNames(
+                styles.footerIcon,
+                isOwnArticle ? styles.disable : styles.reaction,
+                isLike && styles.pressed,
+              )}
+              label={String(likeCount)}
+              onClick={(): void => handleReaction(true)}
```
Comment (andanf-e) [inline frontend/src/pages/articles/components/article-card/article-card.tsx]:
```suggestion
              onClick={handleLikeReaction}
```

File: frontend/src/pages/articles/components/article-card/article-card.tsx [hunk 5]
```diff
@@ -91,21 +119,40 @@ const ArticleCard: React.FC<Properties> = ({
...
+          <li>
+            <IconButton
+              iconName="dislike"
+              className={getValidClassNames(
+                styles.footerIcon,
+                isOwnArticle ? styles.disable : styles.reaction,
+                isLike === false && styles.pressed,
+              )}
+              label={String(dislikeCount)}
+              onClick={(): void => handleReaction(false)}
```
Comment (andanf-e) [inline frontend/src/pages/articles/components/article-card/article-card.tsx]: Create separate handlers that would call generic handler
eg
```
const handleDislikeReaction = () => {
  handleReaction(false);
};
```
```suggestion
              onClick={handleDislikeReaction}
```

File: frontend/src/pages/articles/components/article-card/article-card.tsx [hunk 6]
```diff
@@ -30,19 +48,29 @@ const ArticleCard: React.FC<Properties> = ({
...
 
   const isMyArticles = matchPath(
     { path: `${AppRoute.ARTICLES}/${ArticleSubRoute.MY_ARTICLES}` },
     pathname,
   );
-  const { publishedAt, title, text, id, coverUrl } = article;
-  const { comments, views, likes, dislikes } = reactions;
+  const { publishedAt, title, text, id, userId, coverUrl } = article;
+  const MOCKED_READ_TIME = '7 min read';
+  const { likeCount, dislikeCount, isLike } = getReactionsInfo(
```
Comment (andanf-e) [inline frontend/src/pages/articles/components/article-card/article-card.tsx]: Not sure if this `isLike` is not conflicting with param in handleReaction method

File: frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import { type ReactionResponseDto } from '~/packages/articles/articles.js';
+
+type ReturnValue = {
+  isLike?: boolean;
+  likeCount: number;
+  dislikeCount: number;
+};
```
Comment (andanf-e) [inline frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts]:
```suggestion
type ReactionsInfo = {
  hasAlreadyReactedWith: 'like' | 'dislike' | null;
  likesCount: number;
  dislikesCount: number;
};
```

File: frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    reaction.isLike ? likeCount++ : dislikeCount++;
+  }
+
+  return {
+    isLike,
+    likeCount,
+    dislikeCount,
+  };
```
Comment (andanf-e) [inline frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts]:
```suggestion
  return reactions.reduce<ReactionsInfo>((reactionsInfo, reaction) => {
      const reactionType: ReactionsInfo['hasAlreadyReactedWith'] = reaction.isLike
        ? 'like'
        : 'dislike';

      return {
        likesCount: reaction.isLike
          ? reactionsInfo.likesCount + 1
          : reactionsInfo.likesCount,
        dislikesCount: reaction.isLike
          ? reactionsInfo.dislikesCount
          : reactionsInfo.dislikesCount + 1,
        hasAlreadyReactedWith: reaction.userId === userId
          ? reactionType
          : reactionsInfo.hasAlreadyReactedWith
      };
    },
    { likesCount: 0, dislikesCount: 0, hasAlreadyReactedWith: null }
  );
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/slices/articles/articles.slice.ts (hunk 1 above)
Comment (hnoievets) [reply frontend/src/slices/articles/articles.slice.ts]: Would it be better to add a property `articleReactionDataStatus` or change `dataStatus` like this?
```typescript
dataStatus = {
  articleReaction: dataStatusType,
  ...
};
```

Comment (andanf-e) [reply frontend/src/slices/articles/articles.slice.ts]: Let's add separate property `articleReactionDataStatus` for now

File: frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts (hunk 8 above)
Comment (hnoievets) [reply frontend/src/libs/helpers/get-reactions-info/get-reactions-info.helper.ts]: Got it, but my version was simpler)

File: frontend/src/pages/articles/components/article-card/styles.module.scss [hunk 9]
```diff
@@ -134,19 +134,34 @@
...
   color: var(--dark-gray-50);
   font-weight: 400;
   font-size: 12px;
   font-family: Lato, sans-serif;
   font-style: normal;
   line-height: normal;
+  background-color: var(--white);
+}
+
+.disable {
```
Comment (artemmatiushenko1) [inline frontend/src/pages/articles/components/article-card/styles.module.scss]: `disabled`

File: frontend/src/assets/img/icons/dislike-icon.svg [hunk 10]
```diff
@@ -1,6 +1,6 @@
...
...9 7.17113 0.616254 7.75288C1.11459 8.33354 1.84445 8.66667 2.61786 8.66667H5.11836C4.96235 9.07833 4.80635 9.6785 4.80635 10.4271C4.80635 12.4621 5.97831 13 6.59793 13C7.74233 13 7.83825 11.9513 7.83825 10.9688C7.83825 9.7695 9.20702 8.77283 9.92971 8.32975C9.96665 8.81942 10.3795 9.20833 10.8872 9.20833H12.2654C12.7973 9.20833 13.2301 8.78313 13.2301 8.26042V0.947917C13.2301 0.425208 12.7973 0 12.2654 0ZM7.01137 10.9688C7.01137 12.1875 6.80134 12.1875 6.59793 12.1875C6.20984 12.1875 5.63323 11.7184 5.63323 10.4271C5.63323 9.22675 6.11172 8.49117 6.11503 8.48575C6.19937 8.36117 6.20764 8.20083 6.13542 8.06867C6.06376 7.9365 5.92319 7.85417 5.77104 7.85417H2.61786C2.08866 7.85417 1.58977 7.62667 1.24909 7.22963C0.907868 6.83204 0.763439 6.30879 0.853294 5.79475L1.4707 2.27392C1.61899 1.42675 2.36098 0.8125 3.23527 0.8125H4.11728C7.86747 0.8125 9.74615 1.96517 9.92255 2.07946V7.39321C9.46335 7.63263 7.01137 9.00304 7.01137 10.9688ZM12.4032 8.26042C12.4032 8.33517 12.3415 8.39583 12.2654 8.39583H10.8872C10.8112 8.39583 10.7494 8.33517 10.7494 8.26042V0.947917C10.7494 0.87425 10.8123 0.8125 10.8872 0.8125H12.2654C12.3415 0.8125 12.4032 0.873167 12.4032 0.947917V8.26042Z" fill='black'/>
```
Comment (artemmatiushenko1) [inline frontend/src/assets/img/icons/dislike-icon.svg]: Try to set `fill='currentColor`' instead, and then you can change the SVG's color by setting the `color` CSS property on a parent element, or SVG itself. I guess then there won't be any need to put `path` in selector here `frontend/src/pages/articles/components/article-card/styles.module.scss`

File: frontend/src/slices/articles/actions.ts (hunk 2 above)
Comment (hnoievets) [reply frontend/src/slices/articles/actions.ts]: In this case, the BE needs to respond with something other than `ArticleReactionsResposeDto` after deletion on updating

File: frontend/src/assets/img/icons/dislike-icon.svg (hunk 10 above)
Comment (hnoievets) [reply frontend/src/assets/img/icons/dislike-icon.svg]: Thank you

File: frontend/src/libs/helpers/get-reaction-converted-to-boolean/get-reaction-converted-to-boolean.helper.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import { Reaction } from '~/libs/enums/enums.js';
+import { type ValueOf } from '~/libs/types/types.js';
+
+const getReactionConvertedToBoolean = (
```
Comment (andanf-e) [inline frontend/src/libs/helpers/get-reaction-converted-to-boolean/get-reaction-converted-to-boolean.helper.ts]: Not sure if it should be common helper - as it's used only in one place for article

File: frontend/src/libs/helpers/get-reaction-converted-to-boolean/get-reaction-converted-to-boolean.helper.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  reaction: ValueOf<typeof Reaction>,
+): boolean => {
+  return reaction === Reaction.LIKE ? true : false;
```
Comment (andanf-e) [inline frontend/src/libs/helpers/get-reaction-converted-to-boolean/get-reaction-converted-to-boolean.helper.ts]:
```suggestion
  return reaction === Reaction.LIKE;
```

File: frontend/src/slices/articles/articles.slice.ts [hunk 13]
```diff
@@ -54,13 +58,59 @@ const { reducer, actions, name } = createSlice({
...
+        );
+
+        if (reactionIndex === -1) {
+          article.reactions.push(updatedReaction);
+          return article;
+        }
+
+        article.reactions[reactionIndex] = updatedReaction;
+
+        return article;
```
Comment (andanf-e) [inline frontend/src/slices/articles/articles.slice.ts]:
```suggestion
        const existingReaction = article.reactions.find(
          ({ id }) => id === updatedReaction.id,
        );

        if (!existingReaction) {
          return {
            ...article,
            reactions: article.reactions.concat(updatedReaction);
          };
        }

        const reactionsToUpdate = article.reactions.slice();
        reactionsToUpdate[reactionIndex] = updatedReaction;

        return {
          ...article,
          reactions: reactionsToUpdate
        };
```

File: frontend/src/slices/articles/articles.slice.ts [hunk 14]
```diff
@@ -54,13 +58,59 @@ const { reducer, actions, name } = createSlice({
...
+    builder.addCase(deleteArticleReaction.fulfilled, (state, action) => {
+      const { articleId, reactionId } = action.payload;
+
+      state.articles = state.articles.map((article) => {
+        if (article.id === articleId) {
+          const reactionIndex = article.reactions.findIndex(
+            ({ id }) => id === reactionId,
+          );
+
+          article.reactions.splice(reactionIndex, 1);
```
Comment (andanf-e) [inline frontend/src/slices/articles/articles.slice.ts]:
```suggestion
          const reactionsToUpdate = article.reactions.slice();
          
          reactionsToUpdate.splice(reactionIndex, 1);

          return {
            ...article,
            reactions: reactionsToUpdate
          };
```

**Review by andanf-e — CHANGES_REQUESTED** [review]

File: frontend/src/slices/articles/articles.slice.ts (hunk 14 above)
Comment (hnoievets) [reply frontend/src/slices/articles/articles.slice.ts]: I thought that we can change value directly

File: frontend/src/slices/articles/articles.slice.ts (hunk 13 above)
Comment (hnoievets) [reply frontend/src/slices/articles/articles.slice.ts]: I've changed it, but using the spread operator because eslint doesn't allow `concat` and `slice()`

**Review by andanf-e — APPROVED** [review]

**Review by mhrechkin — APPROVED** [review]

**Pull request merged**